packages/gatsby-transformer-mdx-introspection/gatsby-node.js

```diff
@@ -1,89 +1,355 @@
+  effect(node);
+  if (typeof node === 'object' && 'children' in node) {
+    node.children.forEach(child => visit(child, effect));
+  }
+}
+
+/**
+ * Utility method to implement common query functionality between
+ * childrenComponentInMdx and childComponentInMdx
+ * @param {object} node Gatsby ComponentInMdx node
+ * @param {object} context Gatsby API helpers context
+ */
+async function queryChildren(node, context, { filter, sort, deep, first }) {
+  let baseFilter = {};
+
+  if (deep) {
+    // Traverse data model to find all children IDs
+    const allDeepChildrenIds = [];
+    node.tree.forEach(n =>
+      visit(n, child => {
+        if (typeof child === 'object') {
+          allDeepChildrenIds.push(child.id);
+        }
+      })
+    );
+    baseFilter = { id: { in: allDeepChildrenIds } };
+  } else {
+    // Directly query children
+    baseFilter = { parent: { id: { eq: node.id } } };
+  }
+
+  // Perform query
+  const derivedFilter = merge(filter, baseFilter);
+  const result = await context.nodeModel.runQuery({
+    query: { filter: derivedFilter, sort },
+    type: nodeName,
+    firstOnly: first,
+  });
+  return result;
+}
+
+exports.createSchemaCustomization = ({ actions }) => {
+  actions.createTypes(
+    `"""
+     JSX element attribute with left-hand identifier and right-hand value
+     """
+     type Attribute {
+       "Attribute left-hand identifier (or shorthand identifier)"
+       name: String!
+       "Attribute right-hand value (or JSX expression snippet)"
+       value: JSON
+     }
+
+     """
+     Single JSX element that was contained in a compiled MDX file's output
+     """
+     type ${nodeName} implements Node @childOf(types: ["${nodeName}", "Mdx"], many: true) @dontInfer {
+       "JSX component/tag/element name"
+       component: String!
+       "Original JSX snippet from compiled MDX"
+       jsx: String!
+       "Array of attribute objects describing each parseable attribute specified"
+       attributes: [Attribute!]!
+       "Whether the component is a root node of the MDX document (or of its own detached subtree)"
+       isRoot: Boolean!
+       "Whether the component is from a component tree that wasn't in the direct document tree"
+       isDetached: Boolean!
+       "Array of component nodes that are children of the current node"
+       tree: [JSON!]!
+       "Original MDX file parent node"
+       mdx: Mdx! @link
+       "Text content of the current node. To configure text processing rules, use collapse/trim"
+       content(collapse: Boolean, trim: Boolean): String!
+       "A single child JSX/HTML node (excluding text children). If deep is set, then the field could take the value of any child of children of this node at any level"
+       child${nodeName}: ${nodeName}
+       "All child JSX/HTML nodes (excluding text children). If deep is set, then the field will return all children of children of this node at any level"
+       children${nodeName}: [${nodeName}]
```

Comment:
Yes, I believe so, but I think I tried it before and it interacted strangely with Gatsby's auto-generated children node. I'll look into it

Reply:
I just tested it out and it seems like it does replace the non-null assertions with it's own return type. On the query info screen in GraphiQL, it shows this (also without the description):
![example](https://imgur.com/YVpVUnc.jpg)
I did add the assertions in, but they don't affect the final schema because of Gatsby's behavior. I remember wrestling with this for a while, but I'm not sure how to disable it and remove all warnings. Either way, I think the current state is sufficient, as not having the assertions/description are the only things that are wrong with the way it is currently.